FR #956: Provider-calculated split cost allocation (#1394)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/cost_and_usage/columns/allocatedmethoddetails.md

Reviewer: I made a few organizational changes to #1350 per #1465.  I would expect this PR to follow that same flow.

Please review #1465 apply that same organization here.  If you disagree, please leave some feedback either here, or in #1350, describing why.

Reviewer 2: After review, it seems this PR already conforms to the same separation/organization as the PRs you linked to. If you disagree, please submit a change suggestion.

  * Objects inside "Elements" MUST contain key-value pairs (allocation properties).
    * FOCUS-defined allocation properties adhere to the following additional requirements:
      * Allocation property key MUST match the spelling and casing specified for the FOCUS-defined property.
      * Allocation property value MUST be of the type specified for that property.
      * Allocation property MUST adhere to additional normative requirements specific to that property.
    * Provider-defined allocation properties MAY be included in "Elements".
      * Allocation property keys MUST begin with the string "x_" unless it is a FOCUS-defined allocation property.
* AllocatedMethodDetails root object MAY contain additional provider-defined items, in addition to "Elements".

### Content Requirements

The following keys are used for allocation properties to facilitate querying data across allocations and across providers. Focus-defined keys will appear in the list below and Provider-defined keys will be prefixed with "x_" to make them easy to identify as well as prevent collisions.

<b>Allocated Ratio</b>

> Allocated Ratio communicates the percentage of the Origin Record that this Allocated Record derived from the corresponding [Allocated Method Id](#allocatedmethodid) and Usage Unit property.

Reviewer 3: Does the Allocated Record represent a **Split Record** or a specific **Element** within the array? Allocated Ratio communicates the percentage that the element derived, right? If the Allocated Record represents the Split Record, then this description may be incorrect.

In addition, this definition references the Usage Unit. In the case of idle resources or unallocated portions, I’d expect a single element with AllocatedRatio not null and != 0 if some percentage of cost (and/or quantity) metrics is allocated to this element (and thus the split record). However, we need to clarify whether UsageQuantity (and UsageUnit) are also mandatory and not-nullable for each element.

At this point, I’m just pointing out that we should account for these cases. Unfortunately, I don’t have an alternative proposal ready. 😞

Author: Good call. Note to self: I need to update verbiage to Allocated Charge.
https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/f7b2c87c07a3ffc94b01f72a20afd790249cb441/specification/glossary.md?plain=1#L7

Author: Updated verbiage, linked to the Glossary, and clarified that it is the sum of allocated of values for all AllocatedRatio properties across all allocated charge records related to an origin charge that needs to add up to 1.

Unallocated portions may be included in their own records or may be included as part of the usage records.
https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/11108ae6882b06d3deb426f594a5a6b3c98a8bb7/specification/attributes/provider_calculated_split_costallocation_handling.md?plain=1#L20
https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/11108ae6882b06d3deb426f594a5a6b3c98a8bb7/specification/attributes/provider_calculated_split_costallocation_handling.md?plain=1#L21

# Allocated Method Details

Allocated Method Details provides information about how resources are allocated when usage records are split to support cost allocation requirements.

Allocated Resource Details consists of a valid JSON object which contains an array consisting of key-value objects describing the one or more factors that determined the split cost allocation. Each object consists of FOCUS-defined keys but can be extended to provide additional details about the allocation.

The FOCUS-defined properties are:

* `Allocated Ratio`: The ratio of a [*charge*](#glossary:charge) that this allocation represents.
* `Usage Unit`: Unit being measured used to calculate this allocation.
* `Usage Quantity`: The value of the charge applied to a single contract term.

In addition to these, a provider may include one or more custom properties, also denoted as key-value pairs.

## Requirements

### Column Requirements

The AllocatedMethodDetails column adheres to the following requirements:

> * AllocatedMethodDetails SHOULD be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports [Provider-Calculated Split Cost Allocation](#providercalculatedsplitcosthandling).

Reviewer: This will link to the supported feature, not the attribute.

```suggestion
* AllocatedMethodDetails SHOULD be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports [Provider-Calculated Split Cost Allocation](#provider-calculatedsplitcostallocation).
```

Reviewer 2: Updated outside of this comment because it became outdated.

* AllocatedMethodDetails MUST be of type String.
* AllocatedMethodDetails MUST conform to [StringHandling](#stringhandling) requirements.

> * AllocatedMethodDetails MUST conform to [ObjectFormat](#objectformat) requirements.

Reviewer: ```suggestion
* AllocatedMethodDetails MUST conform to [JsonObjectFormat](#jsonobjectformat) requirements.
```

>   * AllocatedMethodDetails MUST conform to ObjectFormat schema requirements below.

Reviewer: Is this line necessary...?

Reviewer 2: Considering this line in the JSON Object attribute:

> * JsonObjectFormat columns MUST conform to all requirements of the corresponding column definition, which may specify or restrict the shape or contents of the Object.

I'm good with striking it if @kk09v is.

# Allocated Method Details

Allocated Method Details provides information about how resources are allocated when usage records are split to support cost allocation requirements.

Allocated Resource Details consists of a valid JSON object which contains an array consisting of key-value objects describing the one or more factors that determined the split cost allocation. Each object consists of FOCUS-defined keys but can be extended to provide additional details about the allocation.

The FOCUS-defined properties are:

* `Allocated Ratio`: The ratio of a [*charge*](#glossary:charge) that this allocation represents.
* `Usage Unit`: Unit being measured used to calculate this allocation.
* `Usage Quantity`: The value of the charge applied to a single contract term.

In addition to these, a provider may include one or more custom properties, also denoted as key-value pairs.

## Requirements

### Column Requirements

The AllocatedMethodDetails column adheres to the following requirements:

> * AllocatedMethodDetails SHOULD be present in a cost and usage [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports [Provider-Calculated Spli Cost Allocation](#ProviderCalculatedSplitCostAllocation).

Reviewer: ```suggestion
* AllocatedMethodDetails SHOULD be present in a cost and usage [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports [Provider-Calculated Split Cost Allocation](#ProviderCalculatedSplitCostAllocation).
```

### Scenario 1: Single "UsageUnit" value used for allocation

When only a single "UsageUnit" is used to calculate the allaction.

```json
{
  "Elements" : [ {
    "AllocatedRatio" : 0.1,
    "UsageUnit" : "Hours",
    "UsageQuantity" : 300
    }
  ]
}
```
### Scenario 2: Multiple "UsageUnit" values used for allocation

When multiple "UsageUnit" values are used to calculate the allaction, another object is added to the "Elements" collection.

```json
{
  "Elements": [
    {
      "AllocatedRatio": 0.05,
      "UsageUnit": "CPU",
      "UsageQuantity": 0.5
    },
    {
      "AllocatedRatio": 0.1,
      "UsageUnit": "Memory",
      "UsageQuantity": 4
    }
  ]
}
```
### Scenario 3: Provider omits keys that are not required

This provider does not wish to supply the "UsageUnit" or "UsageQuantity" keys but still provides cost allocation with some additional allocation method details. In this case, "UsageUnit" and "UsageQuantity" are omitted, and only the "AllocatedRatio" is supplied.

```json
{
  "Elements" : [ {
    "AllocatedRatio" : 0.45
    }
  ]
}
```

> ### Scenario 3: Additional non-FOCUS specified properties

Reviewer: ```suggestion
### Scenario 4: Additional non-FOCUS specified properties
```

A provider can add additional properties if they feel more context is helpful or neccessary to the practitioner. In this scenario,the provider is supplying additional context that shows only 0.5 of a unit was used. However, since 1 unit was requested by the service this allocation represents, the allocation is being charged at 1 regardless.

```json
{
  "Elements": [
    {
      "AllocatedRatio": 0.6,
      "UsageUnit": "vCPU",
      "UsageQuantity": 1,
      "x_ReservedVCPU": 1,
      "x_UsedVCPU": 0.5,
      "x_AllocatedVCPU": 1
    }
  ]
}
```
## Column ID

AllocatedMethodDetails

## Display Name

Allocated Method Details

## Description

> Allocated Method Details provides information about how resources are allocated when using split cost allocation.

Reviewer: ```suggestion
A set of properties describing how resources are allocated in provider-defined split cost allocation.
```

## Content Constraints

| Constraint      | Value           |
|:----------------|:----------------|
| Column type     | Dimension       |

> | Feature level   | Conditional     |

Reviewer: Because of the SHOULD normative requirement for column presence, this needs to be `Recommended` instead of `Conditional`.
```suggestion
| Feature level   | Recommended     |
```
